core/unix/injector.c

```diff
@@ -1222,24 +1293,40 @@ injectee_run_get_retval(dr_inject_info_t *info, void *dc, instrlist_t *ilist)
      * execution, tracee PC will be set back to syscall instruction
      * PC = PC - sizeof(syscall). We have to add offsets to compensate.
      */
-    if (!info->wait_syscall) {
-        uint offset = 0;
-#    ifdef X86
-        offset = SYSCALL_LENGTH;
+    uint offset = 0;
+    if (!info->wait_syscall)
+        offset = system_call_length(app_mode);
+#    ifdef AARCH64
+    /* POKEUSER is not available. */
+    ptr_int_t saved_pc = regs.REG_PC_FIELD;
+    regs.REG_PC_FIELD = (ptr_int_t)(pc + offset);
+    r = our_ptrace_setregs(info->pid, &regs);
+    if (r < 0)
+        return r;
+    regs.REG_PC_FIELD = saved_pc;
```

Comment:
nit: If this line is not part of the if/else block, move it up before `if` or down after the `else`?

Reply:
This is part of this `if`, restoring the app's pc for the post-execution restore at the end of the function.